server/src/services/library.service.ts

```diff
@@ -570,100 +591,78 @@ export class LibraryService implements OnEvents {
-      await this.assetRepository.updateAll(assetIdsToMarkOffline, { isOffline: true });
-    }
+    let crawlDone = false;
+    let crawlCounter = 0;
+    let crawledAssetPaths: string[] = [];
 
-    if (assetIdsToMarkOnline.length > 0) {
-      this.logger.debug(`Found ${assetIdsToMarkOnline.length} online asset(s) previously marked as offline`);
-      await this.assetRepository.updateAll(assetIdsToMarkOnline, { isOffline: false });
-    }
+    // (Re-)import all assets found on disk
+    while (!crawlDone) {
+      const assetGenerator = await crawledAssets.next();
+      crawlDone = assetGenerator.done ?? true;
 
-    if (crawledAssetPaths.size > 0) {
-      if (!shouldScanAll) {
-        this.logger.debug(`Will import ${crawledAssetPaths.size} new asset(s)`);
+      if (!crawlDone) {
+        crawledAssetPaths.push(assetGenerator.value);
+        crawlCounter++;
       }
 
-      let batch = [];
-      for (const assetPath of crawledAssetPaths) {
-        batch.push(assetPath);
+      if (crawledAssetPaths.length % LIBRARY_SCAN_BATCH_SIZE === 0 || crawlDone) {
+        // We have reached the batch size or the end of the generator, scan the batch
+        this.logger.log(`Queueing scan of ${crawledAssetPaths.length} crawled asset(s) in library ${library.id}...`);
 
-        if (batch.length >= LIBRARY_SCAN_BATCH_SIZE) {
-          await this.scanAssets(job.id, batch, library.ownerId, job.refreshAllFiles ?? false);
-          batch = [];
-        }
-      }
-
-      if (batch.length > 0) {
-        await this.scanAssets(job.id, batch, library.ownerId, job.refreshAllFiles ?? false);
+        await this.scanAssets(job.id, crawledAssetPaths, library.ownerId, job.refreshAllFiles ?? false);
+        crawlCounter += crawledAssetPaths.length;
+        crawledAssetPaths = [];
       }
     }
 
-    await this.repository.update({ id: job.id, refreshedAt: new Date() });
-
-    return JobStatus.SUCCESS;
-  }
-
-  private async getPathTrie(library: LibraryEntity): Promise<Trie<string>> {
-    const pathValidation = await Promise.all(
-      library.importPaths.map(async (importPath) => await this.validateImportPath(importPath)),
+    let existingAssetsDone = false;
+    let existingAssetCounter = 0;
+    const onlineAssets = usePagination(LIBRARY_SCAN_BATCH_SIZE, (pagination) =>
+      this.assetRepository.getWith(pagination, WithProperty.IS_ONLINE, job.id),
     );
 
-    const validImportPaths = pathValidation
-      .map((validation) => {
-        if (!validation.isValid) {
-          this.logger.error(`Skipping invalid import path: ${validation.importPath}. Reason: ${validation.message}`);
-        }
-        return validation;
-      })
-      .filter((validation) => validation.isValid)
-      .map((validation) => validation.importPath);
+    // Check all existing online assets to see if they are still online
+    while (!existingAssetsDone) {
```

Comment:
change this for a for loop. for examples, just look at other places in the code `usePagination` is used.

Reply:
I think what you actually want is what I already commented [here](https://github.com/immich-app/immich/pull/7934#discussion_r1553707433), so no need to even look for examples :P

Reply:
@jrasm91 I've looked into using pagination instead of the current approach which would be idea. However, in this case we have a generator (which internally uses a ReadableStream). Would you be fine with me modifying usePagination to consume generators, too?

Reply:
Done. I implemented the batching in the walk* function directly